packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts

```diff
@@ -0,0 +1,406 @@
+      ) {
+        if (userProfile) {
+          if (userProfile.email) return userProfile.email;
+          else return 'userProfile email is undefined';
+        } //if
+        else return 'userProfile is undefined';
+      }
+    }
+
+    app.controller(InfoController);
+
+    const expiredToken =
+      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjEiLCJlbWFpbCI6ImpvZUBleGFtcGxlLmNvbSIsIm5hbWUiOiJqb2Ugam9lbWFuIiwiaWF0IjoxNTU1ODY3NDAzLCJleHAiOjE1NTU4Njc0NjN9.QKmO5qDC8Yg-aK3EedLRsXczL7VQDDnWtA-cpyqszqM';
+
+    await whenIMakeRequestTo(server)
+      .get('/whoAmI')
+      .set('Authorization', 'Bearer ' + expiredToken)
+      .expect(401);
+  });
+
+  it('returns error due to invalid token #1', async () => {
+    class InfoController {
+      constructor() {}
+
+      @get('/whoAmI')
+      @authenticate('jwt')
+      whoAmI(
+        @inject(AuthenticationBindings.CURRENT_USER) userProfile: UserProfile,
+      ) {
+        if (userProfile) {
+          if (userProfile.email) return userProfile.email;
+          else return 'userProfile email is undefined';
+        } //if
+        else return 'userProfile is undefined';
+      }
+    }
+
+    app.controller(InfoController);
+
+    const invalidToken = 'aaa.bbb.ccc';
+
+    await whenIMakeRequestTo(server)
+      .get('/whoAmI')
+      .set('Authorization', 'Bearer ' + invalidToken)
+      .expect(401);
+  });
+
+  it('returns error due to invalid token #2', async () => {
+    class InfoController {
+      constructor() {}
+
+      @get('/whoAmI')
+      @authenticate('jwt')
+      whoAmI(
+        @inject(AuthenticationBindings.CURRENT_USER) userProfile: UserProfile,
+      ) {
+        if (userProfile) {
+          if (userProfile.email) return userProfile.email;
+          else return 'userProfile email is undefined';
+        } //if
+        else return 'userProfile is undefined';
+      }
+    }
+
+    app.controller(InfoController);
+
+    const invalidToken = 'aaa.bbb.ccc.ddd';
+
+    await whenIMakeRequestTo(server)
+      .get('/whoAmI')
+      .set('Authorization', 'Bearer ' + invalidToken)
+      .expect(401);
+  });
+
+  it('create a json web token throws error for missing email', async () => {
+    class InfoController {
+      constructor(
+        @inject(JWTAuthenticationStrategyBindings.TOKEN_SERVICE)
+        public tokenService: JWTService,
+        @inject(USER_REPO)
+        public users: UserRepository,
+      ) {}
+
+      @get('/createtoken')
+      async createToken() {
+        const joeUser = this.users.list['joe@example.com'].user;
+
+        const joeUserProfile = {
+          id: joeUser.id,
+          email: undefined,
+          name: `${joeUser.firstname} ${joeUser.surname}`,
+        };
+
+        return await this.tokenService.generateToken(joeUserProfile);
+      }
+    }
+
+    app.controller(InfoController);
+
+    await whenIMakeRequestTo(server)
+      .get('/createtoken')
+      .expect(401);
```

Comment:
For a lot of these tests we are asserting the error code, but is it possible to also assert error messages sent back? Not a deal breaker, but it'll make the tests more robust.

Reply:
+1

Reply:
@b-admike , I have addressed your comment. I check error code and message in my tests.